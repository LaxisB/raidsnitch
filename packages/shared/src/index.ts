export * as log from './log';
export * as emitter from './emitter';
export * as rpc from './rpc';
export * as utils from './utils';
export * as format from './format';
