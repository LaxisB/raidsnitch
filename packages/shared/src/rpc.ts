import { wrapLog } from './log';

const log = wrapLog('rpc');
let requestCounter = 0;

function generateId() {
  return requestCounter++;
}

/**
 * create a handler function that dispatches incoming RPC requests to it's handlers
 * @param handlers Record of method names to register with their handler functions
 * @param middleware Array of middleware functions to run on each request
 * @param internalCb adapter for the transport layer (e.g. postMessage)
 */
export function setupRpcHandlers(handlers: Record<string, Function>, internalCb: any) {
  return function (request: RpcRequest<unknown>) {
    if (handlers[request.method]) {
      try {
        const res = handlers[request.method](...((request.params as any) ?? []));
        if (request.id == null) {
          // request was 'notification' style, so we don't need to respond
          return;
        }
        if ('then' in res) {
          return res
            .then((result: any) => {
              internalCb(request, { result });
            })
            .catch((error: any) => {
              internalCb(request, { error });
            });
        }
        internalCb(request, { result: res });
      } catch (e) {
        internalCb(request, { error: e });
      }
    } else {
      log.warn('unhandled request', request);
    }
  };
}

type InternalCb<T> = (req: RpcRequest<any>, transferArgs?: number[]) => Promise<T>;

/**
 * wraps a function to be callable via RPC
 */
export function asRpcCall<Fn extends (...args: any) => any>(name: string, internalCb: InternalCb<ReturnType<Fn>>) {
  return async function (...data: Parameters<Fn>): Promise<Awaited<ReturnType<Fn>>> {
    const request = {
      type: 'rpc' as 'rpc',
      method: name,
      params: data,
      id: generateId(),
    };

    return internalCb(request) as any;
  };
}

/**
 * create a RPC client that can be used to call methods on the server
 * it is a naive implementation that assumes that the server will respond to all requests
 */
export function createNaiveClient<T, B extends {}>(baseObject: B, worker: Worker) {
  const pendingRpcCalls = new Map<number, { resolve: (value: any) => void; reject: (reason?: any) => void }>();
  const post: any = (data: any) => {
    return new Promise((resolve, reject) => {
      log.debug('request', data.id, data.method, data.params);
      pendingRpcCalls.set(data.id, { resolve, reject });
      worker.postMessage(data);
    });
  };

  worker.addEventListener('message', (e) => {
    const res = e.data as RpcResponse<any, any>;
    if (res.type !== 'rpc') {
      return;
    }
    if (pendingRpcCalls.has(res.id)) {
      const { resolve, reject } = pendingRpcCalls.get(res.id)!;
      pendingRpcCalls.delete(res.id);
      if ('error' in res) {
        log.warn('error response', res);
        reject(res.error);
      } else {
        log.debug('response', res.id, res.result);
        resolve(res.result);
      }
    }
  });

  let fnCache: Record<string | symbol, any> = {};
  const client = new Proxy(baseObject, {
    get: function (target, prop, _receiver) {
      if (prop in target) {
        return (target as any)[prop];
      }
      if (fnCache[prop]) {
        return fnCache[prop];
      }
      fnCache[prop] = asRpcCall(prop as string, post);
      return fnCache[prop];
    },
  });
  return client as B & AsyncClient<T>;
}

export type AsyncClient<T> = {
  [K in keyof T]: T[K] extends (...args: infer A) => infer R ? (...args: A) => AsyncResponse<R> : never;
};

type AsyncResponse<T> = T extends Promise<unknown> ? T : Promise<T>;

export interface RpcRequest<T> {
  type: 'rpc';
  method: string;
  params: T;
  id: number;
}

interface RpcResponseBase {
  type: 'rpc';
  id: number;
}
interface RpcErrorResponse<E> extends RpcResponseBase {
  error: {
    message: string;
    data: E;
  };
}
interface RpcSuccessResponse<T> extends RpcResponseBase {
  result: T;
}
export type RpcResponse<T, E> = RpcErrorResponse<E> | RpcSuccessResponse<T>;

export interface RpcEvent {
  type: 'event';
  event: string;
  data: any;
}
