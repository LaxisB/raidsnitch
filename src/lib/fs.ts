import { formatFileSize } from '@/lib/format';
import { wrapLog } from '@/lib/log';
import { sleep } from '@/lib/utils';
import { debug } from './debug';

const log = wrapLog('fileReader');

interface FileReaderOpts {
    /** amount of bytes to read at once
     * @default 64KiB
     */
    chunkSize: number;
    /** keep the file open when reaching the end and watch for changes
     * @default false
     */
    tail: boolean;
    /**change the regex used to split the file into items
     * @default /\n/
     */
    splitRegex: RegExp;
}

/**
 * create a file reader that reads a file in chunks
 * @param handle file handle to read
 * @param options options for the reader
 * @returns a handle to the reader
 * @example
 * const file = await window.showOpenFilePicker();
 * const reader = createFileReader(file, { tail: true }});
 * for await (const res of reader.getReader()) {
 *   if (!res.length) {
 *     // no changes, wait for more
 *     // this waits using an exponential backoff
 *     // we need to do this to avoid running into timeouts regarding await
 *     break;
 *   }
 *   console.log(res);
 * }
 * reader.stop();
 *
 */
export function createFileReader(handle: FileSystemFileHandle, options?: Partial<FileReaderOpts>) {
    let opts: FileReaderOpts = { tail: false, chunkSize: 64 * 1024, splitRegex: /\n/, ...options };
    let doStop = false;
    let reader: AsyncGenerator<string[]> | null = null;

    async function* getReader() {
        let file = await handle.getFile();
        let offset = 0;
        let decoder = new TextDecoder();
        let remainder = '';
        let failedRead = 0;

        while (true) {
            if (doStop) {
                return [];
            }
            // if we reach the end of the current file handle, reload and attempt to read more
            if (offset >= file.size) {
                if (!opts.tail) {
                    return [];
                }
                file = await handle.getFile();
                if (offset >= file.size) {
                    const sleepTime = getBackOffTime(Math.min(5, failedRead++));
                    yield [];
                    await sleep(sleepTime);
                    continue;
                }
            }

            const value = file.slice(offset, offset + opts.chunkSize);
            const buffer = await value.arrayBuffer();
            offset += value.size;
            const chunk = decoder.decode(buffer, { stream: true });
            const lines = (remainder + chunk).split(opts.splitRegex);
            remainder = lines.pop() ?? '';

            debug({
                file: file.name,
                size: file.size,
                offset,
                chunkSize: opts.chunkSize,
                lines: lines.length,
            });
            yield lines;
        }
    }

    return {
        items() {
            reader = reader ?? getReader();
            return reader;
        },
        stop() {
            doStop = true;
        },
    };
}

export type FileReader = ReturnType<typeof createFileReader>;

interface DirWatcherOpts {
    /**
     * emit the newest file on start
     * @default true
     */
    emitOnStart: boolean;
}

/** watch a directory for changes, specifically for new combat logs
 * @param handle directory handle to watch
 * @param options options for the watcher
 * @returns a handle to the watcher
 * @example
 * const dir = await window.showDirectoryPicker();
 * const watcher = createDirWatcher(dir);
 * watcher.onFileChange(async (handle) => {
 *    const file = await handle.getFile();
 *   console.log('newest log changed', file.name);
 * });
 * watcher.start();
 */
export function createDirWatcher(handle: FileSystemDirectoryHandle, options?: Partial<DirWatcherOpts>) {
    let opts: DirWatcherOpts = { emitOnStart: true, ...options };
    let lastActiveFile: FileSystemFileHandle | null = null;
    let onFileChange: ((file: FileSystemFileHandle) => void) | null = null;
    let runningTimeout: NodeJS.Timeout | null = null;

    async function watchForFileChanges() {
        const file = await getLatestCombatLog();
        if (file) {
            if (file.name !== lastActiveFile?.name) {
                lastActiveFile = file;
                debug({
                    dir: handle.name,
                    fileChange: new Date().toLocaleTimeString(),
                });
                onFileChange?.(file);
            }
        }
        runningTimeout = setTimeout(watchForFileChanges, 1000);
    }
    async function getLatestCombatLog() {
        let latest: FileSystemFileHandle | null = null;
        let latestMod = 0;
        for await (let entry of handle.values()) {
            if (!fileIsCombatLog(entry.name)) {
                continue;
            }
            const mod = (await entry.getFile()).lastModified;
            if (mod > latestMod) {
                latest = entry;
                latestMod = mod;
            }
        }
        return latest;
    }

    return {
        onFileChange(cb: (file: FileSystemFileHandle) => void) {
            onFileChange = cb;
        },
        async start() {
            lastActiveFile = await getLatestCombatLog();
            if (opts.emitOnStart && lastActiveFile) {
                onFileChange?.(lastActiveFile);
            }
            watchForFileChanges();
        },
        stop() {
            clearTimeout(runningTimeout!);
        },
    };
}

export type DirWatcher = ReturnType<typeof createDirWatcher>;

function fileIsCombatLog(name: string) {
    return name.startsWith('WoWCombatLog-');
}
function handleIsDirectory(handle: any): handle is FileSystemDirectoryHandle {
    return handle instanceof FileSystemHandle && handle.kind === 'directory';
}

/** exponential backoff fn */
function getBackOffTime(attempt: number) {
    return Math.min(100 * 2 ** attempt, 1000 * 60);
}

(window as any).fileTest = async function () {
    const { createParser } = await import('@/lib/parser');
    const dir = await window.showDirectoryPicker();
    const readerOpts = { replayEvents: false, chunkSize: 1024 * 1024, splitRegex: /^(?=\d)/m };
    let parser = createParser(0);
    let activeReader: ReturnType<typeof createFileReader> | null = null;

    function replaceReader(newReader: ReturnType<typeof createFileReader>) {
        activeReader?.stop();
        activeReader = newReader;

        drainReader(newReader);
    }

    async function drainReader(reader: ReturnType<typeof createFileReader>) {
        log.log('reading file');
        const start = Date.now();
        let total = 0;
        for await (const res of reader.items()) {
            if (!res.length) {
                reader.stop();
                break;
            }
            res.map(parser.parseLine).filter((x) => x !== null);
            total += res.length;
        }
        const end = Date.now();
        console.log(`[${formatFileSize(readerOpts.chunkSize)}]: ${total} events finished after ${end - start}ms`);
    }

    const dirHandler = createDirWatcher(dir, { emitOnStart: true });
    dirHandler.onFileChange(async (handle) => {
        const file = await handle.getFile();
        log.log('newest log changed', file.name);
        const reader = createFileReader(handle, readerOpts);
        parser = createParser(file.lastModified);
        replaceReader(reader);
    });
    dirHandler.start();
};
