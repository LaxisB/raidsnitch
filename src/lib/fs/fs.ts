import { formatFileSize } from '../format';
import { wrapLog } from '../log';
import { sleep } from '../utils';

const log = wrapLog('fileReader');

interface FileReaderOpts {
    replayEvents: boolean;
    // amound of data to read at once
    chunkSize: number;
}
export function createFileReader(handle: FileSystemFileHandle, options?: Partial<FileReaderOpts>) {
    let opts: FileReaderOpts = { replayEvents: false, chunkSize: 1024 * 1024, ...options };
    let doStop = false;
    let reader: AsyncGenerator<string[]> | null = null;

    async function* getReader() {
        let file = await handle.getFile();
        let offset = 0;
        let decoder = new TextDecoder();
        let remainder = '';
        let splitRegex = /^(?=\d)/m;
        let failedRead = 0;

        while (true) {
            if (doStop) {
                return [];
            }
            // if we reach the end of the current file handle, reload and attempt to read more
            if (offset >= file.size) {
                file = await handle.getFile();
                if (offset >= file.size) {
                    const sleepTime = getBackOffTime(Math.min(5, failedRead++));
                    yield [];
                    await sleep(sleepTime);
                    continue;
                }
            }

            const value = file.slice(offset, offset + opts.chunkSize);
            const buffer = await value.arrayBuffer();
            offset += value.size;
            const chunk = decoder.decode(buffer, { stream: true });
            const lines = (remainder + chunk).split(splitRegex);
            remainder = lines.pop() ?? '';
            yield lines;
        }
    }

    return {
        getReader() {
            reader = reader ?? getReader();
            return reader;
        },
        stop() {
            doStop = true;
        },
    };
}

interface DirWatcherOpts {
    emitOnStart: boolean;
}
export function createDirWatcher(handle: FileSystemDirectoryHandle, options?: Partial<DirWatcherOpts>) {
    let opts: DirWatcherOpts = { emitOnStart: false, ...options };
    let lastActiveFile: FileSystemFileHandle | null = null;
    let onFileChange: ((file: FileSystemFileHandle) => void) | null = null;
    let runningTimeout: NodeJS.Timeout | null = null;

    async function watchForFileChanges() {
        if (!handle) {
            return;
        }

        const file = await getLatestCombatLog();
        if (file) {
            if (file.name !== lastActiveFile?.name) {
                lastActiveFile = file;
                onFileChange?.(file);
            }
        }
        runningTimeout = setTimeout(watchForFileChanges, 1000);
    }
    async function getLatestCombatLog() {
        if (!handle) {
            throw new Error('no dirHandle');
        }

        let latest: FileSystemFileHandle | null = null;
        let latestMod = 0;
        for await (let entry of handle.values()) {
            if (handleIsDirectory(entry)) {
                continue;
            }
            if (!fileIsCombatLog(entry.name)) {
                continue;
            }
            const mod = (await entry.getFile()).lastModified;
            if (mod > latestMod) {
                latest = entry;
                latestMod = mod;
            }
        }
        return latest;
    }

    return {
        onFileChange(cb: (file: FileSystemFileHandle) => void) {
            onFileChange = cb;
        },
        async start() {
            if (opts.emitOnStart) {
                lastActiveFile = await getLatestCombatLog();
                if (lastActiveFile) {
                    onFileChange?.(lastActiveFile);
                }
            }
            watchForFileChanges();
        },
        stop() {
            clearTimeout(runningTimeout!);
        },
    };
}

function fileIsCombatLog(name: string) {
    return name.startsWith('WoWCombatLog-');
}
function handleIsDirectory(handle: any): handle is FileSystemDirectoryHandle {
    return handle instanceof FileSystemHandle && handle.kind === 'directory';
}

/** exponential backoff fn */
function getBackOffTime(attempt: number) {
    return Math.min(100 * 2 ** attempt, 1000 * 60);
}

(window as any).fileTest = async function () {
    const { createParser } = await import('@/lib/parser');
    const dir = await window.showDirectoryPicker();
    const readerOpts = { replayEvents: false, chunkSize: 1024 * 1024 };
    let parser = createParser(0);
    let activeReader: ReturnType<typeof createFileReader> | null = null;

    function replaceReader(newReader: ReturnType<typeof createFileReader>) {
        activeReader?.stop();
        activeReader = newReader;

        drainReader(newReader);
    }

    async function drainReader(reader: ReturnType<typeof createFileReader>) {
        log.log('reading file');
        const start = Date.now();
        let total = 0;
        for await (const res of reader.getReader()) {
            if (!res.length) {
                reader.stop();
                break;
            }
            res.map(parser.parseLine).filter((x) => x !== null);
            total += res.length;
        }
        const end = Date.now();
        console.log(`[${formatFileSize(readerOpts.chunkSize)}]: ${total} events finished after ${end - start}ms`);
    }

    const dirHandler = createDirWatcher(dir, { emitOnStart: true });
    dirHandler.onFileChange(async (handle) => {
        const file = await handle.getFile();
        log.log('newest log changed', file.name);
        const reader = createFileReader(handle, readerOpts);
        parser = createParser(file.lastModified);
        replaceReader(reader);
    });
    dirHandler.start();
};
