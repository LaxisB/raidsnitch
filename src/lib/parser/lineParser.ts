const LIST_START = ['[', '('];
const LIST_END = [']', ')'];
const SEPARATORS = [',', ...LIST_START, ...LIST_END];

type LogBottomValues = null | number | string;
type LogValues = LogBottomValues | LogValues[];

export interface DataLine {
    time: 0;
    offset: number;
    event: string;
    payload: LogValues[];
    line: string;
}

export interface ErrorLine {
    time: 0;
    offset: number;
    event: null;
    payload: null;
    line: string;
}

export type LogLine = DataLine | ErrorLine;
export interface Parser {
    parseLine(line: string): LogLine;
}

export function createLineParser(referenceTime: number) {
    let reftime = referenceTime;
    let offset = 0;

    return {
        parseLine(line: string): LogLine {
            const [timeblock, rest] = line.split('  ');
            if (!timeblock || !rest) {
                return {
                    time: 0,
                    offset: offset++,
                    event: null,
                    payload: null,
                    line,
                } as ErrorLine;
            }

            const firstComma = rest.indexOf(',');
            const event = rest.slice(0, firstComma);
            const body = rest.slice(firstComma + 1);

            const time = parseTimeblock(timeblock, reftime);
            const payload = parsePayload(body);

            return {
                time,
                offset: offset++,
                event,
                payload,
                line,
            } as LogLine;
        },
    };
}

const TIMESTAMP_MATCHER = new RegExp(/(\d{1,2})\/(\d{1,2}) (\d{1,2}):(\d{1,2}):(\d{1,2}).(\d{1,3})/);
function parseTimeblock(timeBlock: string, reftime: number): number {
    const matches = TIMESTAMP_MATCHER.exec(timeBlock);
    const d = new Date(reftime);
    d.setMonth(parseInt(matches![1]) - 1);
    d.setDate(parseInt(matches![2]));
    d.setHours(parseInt(matches![3]));
    d.setMinutes(parseInt(matches![4]));
    d.setSeconds(parseInt(matches![5]));
    d.setMilliseconds(parseInt(matches![6]));
    return d.getTime();
}

function parsePayload(body: string) {
    const stack = makeStack();
    const payload = _parsePayload(body, stack);
    return payload as LogValues[];
}

function _parsePayload(rest: string, stack: Stack): any[] {
    // EOL :::: exit branch
    // this is the only
    // we're checking for length to avoid matching long strings
    if (!rest || (rest.length < 2 && /^\s*$/.test(rest))) {
        return stack.value();
    }

    // nil val
    if (rest.startsWith('nil')) {
        stack.addVal(null);
        return _parsePayload(rest.slice(4), stack);
    }

    // flags
    if (rest.startsWith('0x')) {
        const [val, rest2] = splitAtSeparator(rest);
        const flags = parseFlags(val);
        stack.addVal(flags);
        return _parsePayload(rest2, stack);
    }

    // quoted string
    if (rest.startsWith('"')) {
        // assume that the char following the closing quote is a separator
        // there is no precedent to indicate that this is not always the case
        const [val, rest2] = splitAtQuote(rest.slice(1));
        stack.addVal(val.slice(0, -1).trim());
        return _parsePayload(rest2, stack);
    }

    // lists
    if (LIST_START.includes(rest[0])) {
        stack.addLayer();
        return _parsePayload(rest.slice(1), stack);
    }
    if (LIST_END.includes(rest[0])) {
        stack.collapseLayer();
        // remove the separator and trailing comma
        const rest2 = rest.slice(1).replace(/^,/, '');
        return _parsePayload(rest2, stack);
    }

    // special case: a null guid
    if (rest.startsWith('0000000000000000')) {
        const [val, rest2] = splitAtSeparator(rest);
        stack.addVal(val);
        return _parsePayload(rest2, stack);
    }

    // value (int/float/const)
    const [val, rest2] = splitAtSeparator(rest);
    stack.addVal(parsePrimitive(val));
    return _parsePayload(rest2, stack);
}

/**
 * Parses a hex-formatted string of into an integer
 */
function parseFlags(flags: string) {
    if (!flags.startsWith('0x')) {
        return 0;
    }
    return Number.parseInt(flags.slice(2), 16);
}

function parsePrimitive(val: string) {
    const isNum = /^-?\d+(\.\d+)?$/.test(val);
    return isNum ? Number.parseFloat(val) : val;
}

function splitAtQuote(string: string) {
    const offset = string.indexOf('"');
    return [string.slice(0, offset + 1), string.slice(offset + 2)];
}

function splitAtSeparator(string: string) {
    const offsets = SEPARATORS.map((sep) => string.indexOf(sep)).filter((offset) => offset > -1);
    const min = Math.min(string.length, ...offsets);
    const val = string.slice(0, min);
    // trim leading comma if present. we don't need it for parsing
    const rest = string[min] === ',' ? string.slice(min + 1) : string.slice(min);
    return [val, rest];
}

interface Stack {
    addVal(val: any): void;
    addLayer(): void;
    collapseLayer(): void;
    value(): any[];
}

function makeStack(): Stack {
    const values: any[] = [[]];

    const head = () => values[values.length - 1];

    return {
        addVal(val: any) {
            head().push(val);
        },
        addLayer() {
            values.push([]);
        },
        collapseLayer() {
            const head = values.pop();
            values[values.length - 1].push(head);
        },

        value() {
            return values[0];
        },
    };
}
