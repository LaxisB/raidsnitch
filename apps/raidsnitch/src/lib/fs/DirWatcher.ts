import { wrapLog } from '@raidsnitch/shared/log';
import { get, set } from 'idb-keyval';
import { FileHandler, LogStates } from './domain';

const log = wrapLog('dir_watcher');

export class DirWatcher {
    private dirHandle: FileSystemDirectoryHandle | null = null;
    private fileHandle: FileSystemFileHandle | null = null;
    private readDirTimeout: NodeJS.Timeout | null = null;
    protected state = LogStates.INITIAL;

    constructor( private fileHandler: FileHandler) {}

    async restore() {
        const dirHandle = await get('dirHandle');
        if (dirHandle) {
            log.debug('restoring dir handle from idb');
            this.dirHandle = dirHandle;
            this.setState(LogStates.NEED_PERMISSION);
        } else {
            this.setState(LogStates.NEED_DIR);
        }
    }

    async watchDirectory(handle?: FileSystemDirectoryHandle) {
        if (handle) {
            this.dirHandle = handle;
            this.setState(LogStates.NEED_PERMISSION);
            set('dirHandle', handle);
        }
        if (!this.dirHandle) {
            throw new Error('no directory handle');
        }
        const perms = await this.dirHandle.queryPermission();
        if (perms !== 'granted') {
            throw new Error("don't have permission to read directory");
        }

        this.setState(LogStates.HAS_DIR);

        this.watchForFileChanges();
    }

    close() {
        if (this.fileHandler) {
            this.fileHandler.close();
        }
        if (this.readDirTimeout) {
            clearTimeout(this.readDirTimeout);
            this.readDirTimeout = null;
        }
    }

    // poll the directory for the latest interesting file and update the state
    async watchForFileChanges() {
        if (!this.dirHandle) {
            return;
        }

        const file = await this.getLatestCombatLog();

        if (file && file.name != this.fileHandle?.name) {
            this.fileHandle = file;
            this.setState(LogStates.HAS_FILE);
            this.fileHandler.handleFileChange(file);
        }

        // if we're actively reading from a file, we probably don't need to look for a new one
        this.readDirTimeout = setTimeout(
            () => this.watchForFileChanges(),
            this.state === LogStates.HAS_FILE ? 10_000 : 1000,
        );
    }

    private async getLatestCombatLog() {
        if (!this.dirHandle) {
            throw new Error('no dirHandle');
        }

        let latest: FileSystemFileHandle | null = null;
        let latestMod = 0;
        for await (let entry of this.dirHandle.values()) {
            if (handleIsDirectory(entry)) {
                continue;
            }
            if (!fileIsCombatLog(entry.name)) {
                continue;
            }
            const mod = (await entry.getFile()).lastModified;
            if (mod > latestMod) {
                latest = entry;
                latestMod = mod;
            }
        }

        return latest;
    }
    private setState(state: LogStates) {
        this.state = state;
    }
}

export function fileIsCombatLog(name: string) {
    return name.startsWith('WoWCombatLog-');
}
function handleIsDirectory(handle: any): handle is FileSystemDirectoryHandle {
    return handle instanceof FileSystemHandle && handle.kind === 'directory';
}
