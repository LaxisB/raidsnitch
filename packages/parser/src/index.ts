import { parseAsEvent } from './events/parser';
import { createLineParser } from './lineParser';

export type * from './events';
export { casts } from './events';

export type Parser = ReturnType<typeof createParser>;

export function createParser(referenceTime: number) {
  const lineParser = createLineParser(referenceTime);

  function parseLine(line: string) {
    const logLine = lineParser.parseLine(line);
    if (logLine.event === null) {
      return null;
    }
    return parseAsEvent(logLine, referenceTime);
  }

  return {
    parseLine,
  };
}
