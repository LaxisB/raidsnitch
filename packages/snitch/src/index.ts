import { WowEvent } from '@raidsnitch/parser';
import { createDpsHandler } from './handlers/dps';
import { createEntitiesHandler } from './handlers/entities';
import { createHpsHandler } from './handlers/hps';
import { createSegmentHandler } from './handlers/segments';
import { createZoneHandler } from './handlers/zone';

export type State = ReturnType<typeof initialize>['state'];
export interface Snitch {
  state: State;
  handleEvents(events: WowEvent[]): void;
}

export function initialize(handleStats: (stats: State) => void) {
  let refTime = -1;
  const zone = createZoneHandler();
  const entities = createEntitiesHandler();
  const dps = createDpsHandler();
  const hps = createHpsHandler();
  const segments = createSegmentHandler();

  let state = {
    zone: zone.initialState,
    entities: entities.initialState,
    dps: dps.initialState,
    hps: hps.initialState,
    segments: segments.initialState,
  };

  const handlers = [segments.handleEvent, zone.handleEvent, entities.handleEvent, dps.handleEvent, hps.handleEvent];

  function handleEvents(events: WowEvent[]) {
    events.forEach((event) => {
      if (refTime === -1) {
        refTime = event.time;
      }
      handlers.forEach((handler) => handler(event, state));
    });

    handleStats(state);
  }
  return {
    state,
    handleEvents,
  };
}
