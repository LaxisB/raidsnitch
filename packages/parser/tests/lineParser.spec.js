import { readFile } from 'fs/promises';
import { describe, it } from 'vitest';
import { createLineParser } from '../src/lineParser';
const raw = await readFile('D:\\Games\\World of Warcraft\\_retail_\\Logs\\WoWCombatLog-071623_150358.txt', 'utf8');
const lines = raw.split('\n');
describe('lineParser', () => {
    it('should be defined', ({ expect }) => {
        expect(true).toBe(true);
    });
    it('parses data', async ({ expect }) => {
        const parser = createLineParser(0);
        const res = parser.parseLine('7/1 19:58:19.733  COMBAT_LOG_VERSION,20,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,10.1.0,PROJECT_ID,1');
        expect(res).toMatchSnapshot();
    });
    it('does not throw', async ({ expect }) => {
        try {
            const parser = createLineParser(0);
            lines.forEach((line) => {
                parser.parseLine(line);
            });
        }
        catch (e) {
            expect.fail(e.message);
        }
    });
});
