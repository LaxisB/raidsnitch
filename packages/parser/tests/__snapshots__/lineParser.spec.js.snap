// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineParser > parses data 1`] = `
{
  "event": "COMBAT_LOG_VERSION",
  "line": "7/1 19:58:19.733  COMBAT_LOG_VERSION,20,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,10.1.0,PROJECT_ID,1",
  "offset": 0,
  "payload": [
    20,
    "ADVANCED_LOG_ENABLED",
    1,
    "BUILD_VERSION",
    "10.1.0",
    "PROJECT_ID",
    1,
  ],
  "time": 15706699733,
}
`;
